Completed v2 support, added initial v3 support, fixes and clean-up

balyuf opened the pull request:
Hi Andy,

I bundled all my enhancements, fixes and clean-up in one PR commit.  I guess after merging, you will want to go through and adapt things to your coding style, change some selections I made in l9config.py in favor for MSX or add your own ROM versions for the v2 games.

I will try to give an overview of the most important changes:
- Reintroduced support for long descriptions (in 255 byte extends)
- Heuristic for finding the end of the dictionary (and detect duplicates, also fix word start printout)
- Support v1 message database in v2 games (through the mversion variable)
- Autodetect v2 games (for actually identifying them we would need much more)
- Printing of all messages adapted to v2 and cleaned-up
- Printing of all exits adapted to v2 and cleaned-up
- Adapted decompiler.py to v2 (merged level9.py changes, screen handling routines, heuristic for finding end of jump table, heuristic for finding end of A-code)
- Added MSX extracted dat files for v2 games (including Return to Eden)
- Added v2 test-scripts, game texts, exits and decompilation texts

Enjoy!

balyuf committed:
Completed v2 support, fixes, clean-up and initial v3 support

balyuf commented:
I force pushed an amended commit as I had an inspiration for identification of v2 games (through -f file iso -g game) by introducing their full name in the config file and then looking for it in a few key descriptions.  Remark that the signature bytes of v1 would also match the v2 game, so luckily we can autodetect v2 games (through the checksum) and then identify them separately.

balyuf commented:
And another forced push to bring the decompiler to load -f file iso only -g game (and make finding the end of A-code in version 1 more robust again with double 0, which chokes some v2 game too early by coincidence).

balyuf commented:
Well, last little fix through a forced push, to decompiler.py (i.e. the 0x02 end of fragment case) - I'll let you digest all this for the weekend! :-)

balyuf commented:
Through experimenting with v3-4 jump tables, I found a generic solution for all versions.  I could force push that isolated change here to improve v1-2 disassembly, but I don't know where you are with looking at the changes?

balyuf commented:
I've made so many general improvements to decompiler.py in the meantime that I think it does not make sense to isolate the jump tables solution for v1-2 improvement.  Either I force push the whole change to the decompiler (including the preliminary v3-v4) support here, or I wait until you have merged this PR, and then create a second PR.  What do you prefer?

balyuf commented:
Ok, I decided to also force push to this PR my initial v3 support for decompilation.  I hope that is fine with you ...

ajgbarnes commented:
My goal is to write as an authentic as possible specification for each version - I know I could hack together quickly something that works for each version.  I'm decompiling the 6502 games as a reference point (I know that each platform is slightly different in implementation).

I'm going to work through each so I fully understand them all.  I wasted time on the graphics, I should have focused on the text first (lesson learnt).  I would have been here a while ago. 

I'm trying to be authentic to the original code approach too - I know modern ways are more efficient e.g. I see you replace methods versus my looping byte checks.  For me it's an evolution, understand and document and then improve.  

It's all good.  I still have my v2 journey to complete and then v3-4.  And I want to understand the A-code as it reflects what is in exits as an example.

balyuf commented:
I completely understand your wanting to evolve into this.  It's also how I proceed.  So, please, feel free to take and adapt as you see fit.  I think I have given my best this week.  I can wait now ... :-)

ajgbarnes commented:
Thank you Filip - it's sooo important to me because of "being a child" back then and being fascinated about the magic of it all.  I appreciate your help.

I wish I had more free time too :) 

I won't merge this fully into main - I'm picking through the complication(s) and what you did e.g. dictionary.

There is a clearly a dictionary bug that went unnoticed in v2/3 where the dictionary was expected to be terminated with a 0x00 but never was.... so it runs in memory until it finds one. BBC code assumes a 0x00 at the end of the dictionary, but their tools had a bug (love this) but it works anyway because of "garbage".  I wonder if I can abuse this with partial words that match.... and see what the A-code does.

ajgbarnes commented:
It's also part of the "digital archaeology" that these artefacts exist - I think maybe the v3 BBC games made an assumption (I don't know yet) on max dictionary id ? We can brute force it - we are doing unnatural things with the dictionary now when trying to print it :D

balyuf commented:
In my experience, for the dictionary, the minimal thing you need to do, if you stick to building the dictionary into a list, is to check for duplicates.  Of course, if you really want to be close to the original, you should not even do that.  And then the only problem that could be is, that some non-intended word fragments will match, result in an unknown code, and potentially crash the game.

So, though I sympathize with your goal of staying close to the original code, I feel it is hard to reconcile with your goal of being able to run all games in all versions on all platforms.  That being said, how do we proceed?  Currently my master branch is blocked because of the PR.  Would you closing the PR without merging it (so that you still have access to the diffs) help me out?  I think then next pushes to my master branch will not figure in the closed PR, effectively freezing it for your convenience?

With that in view, I could continue to implement v3-4 features, which, if needed at some point in your evolution, you could then have a look at, and adapt to your needs, to speed up your process?  On a philosophical note, even me helping you to speed up can be construed as interference with your evolutionary process.  So, I wonder where the balance lies?  Well, my proposal seems to be a way to discover that balance from moment to moment?

balyuf commented:
PS: checking for duplicates is even necessary when the dictionary would be terminated with a 0, as most games have already legal duplicates.  So, to mimick the original search algorithm, which will stop at first match, you must filter the list to only retain the first occurrence, hence the duplicates check I had introduced.  But, if you want to stay close to the original code, you should not use a Python list of course ... ;-)

balyuf commented:
PPS: that is assuming, that there is no way a search can start in the middle of the dictionary - I have not found any evidence for that, did you?

balyuf commented:
PPPS: so, it keeps me wondering why then there are duplicates, often with different codes?  Do they have a functional purpose, or are they oversights?  You said something that the order of the dictionary is important?  Perhaps there is some hint there?

ajgbarnes commented:
> PPPS: so, it keeps me wondering why then there are duplicates, often with different codes? Do they have a functional purpose, or are they oversights? You said something that the order of the dictionary is important? Perhaps there is some hint there?

They have to be oversights, certainly in any v1 machine code - will just match against the first it finds. 

I haven't checked the v2 machine code for all Level 9 games, just Emerald Isle and Return to Eden and they behave the same way as v1.  It sounds as though they were just left in there in the development cycle.

I don't think it's related to "order", that I have seen so far in the A-code.  The A-code always sorts the words into ascending id order with (from memory but you get the idea) the dictionary being structured:

1. Directions
2. Verbs
3. Objects
4. NPCs
5. Inaminate objects that the player can interact with
6. Other special words like everything or numbers 

The A-code can ALWAYS assume that the first word is a direction or verb (for general input) and if not, print an error (less than a certain value when using jump tables or not matched in a LOT of if statements).

It also tries to makes sense of the player's input so if I type "troll emerald give" it will put it into the order it understands i.e. "give emerald troll" so then it can check to see if there is <direction / verb> only, <direction / verb> <object> or<direction> <verb> <npc> or <direction> <verb> <special> etc

So if the player types "turn cog 6" it will match <direction> <verb> <special> and assumptions made on where to look for <special> e.g. the number and the range it should be in.

The A-code can the be kind (?) on how it processes too - if a player types "N GET LAMP" it will just take the player North.  Same with "GET LAMP N" will re-sort and then go North.

ajgbarnes commented:
> PPS: that is assuming, that there is no way a search can start in the middle of the dictionary - I have not found any evidence for that, did you?

No never seen this - I'll take a look at where the duplicates are later in the week.  I strongly feel they are oversights.

ajgbarnes commented:
> PS: checking for duplicates is even necessary when the dictionary would be terminated with a 0, as most games have already legal duplicates. So, to mimick the original search algorithm, which will stop at first match, you must filter the list to only retain the first occurrence, hence the duplicates check I had introduced. But, if you want to stay close to the original code, you should not use a Python list of course ... ;-)

Haha yes on Python list... but oh the convenience.  And I do think it's a bug they introduced where there is no 0x00 terminator.  We'll never know but the BBC 6502 code (Return to Eden and Emerald Isle) expects it.  Otherwise just goes reading off into garbage memory until it finds one.

ajgbarnes commented:
Yes I can close the PR without merging - I'm going to have to be careful, I think I said I was working on an experimental debugger which I committed today.

balyuf commented:
> Yes I can close the PR without merging - I'm going to have to be careful, I think I said I was working on an experimental debugger which I committed today.

Wow!  The debugger is really a nice enhancement!

I makes it harder though to integrate some of my changes ...  It seems both our efforts went in completely different directions (I don't remember you mentioning working on a debugger though), and are substantial ...  Well, we'll see how things go ...

balyuf commented:
> It makes it harder though to integrate some of my changes ... It seems both our efforts went in completely different directions (I don't remember you mentioning working on a debugger though), and are substantial ... Well, we'll see how things go ...

Ok, I was able to merge in your changes and force push again to this PR.  There were only two slight conflicts, which were easy to resolve ...

balyuf commented:
BTW, there seems to be still a problem with the debugger where after a while the exits cause an index exception ...
EDIT: that is with Lords of Time and Snowball ...
EDIT2: it's the part directions[vm_variables[moveDirectionVar]] which goes out of bounds at direction 16 (max 15 allowed) ... I guess the code above handles it differently than the debug print?

ajgbarnes commented:
Yeah I'm not surprised - I haven't exhaustively tested it but wanted it out there as "experimental".  I think I can guess what's gone wrong 😄  I cannot look tonight.

balyuf commented:
> Yeah I'm not surprised - I haven't exhaustively tested it but wanted it out there as "experimental". I think I can guess what's gone wrong 😄 I cannot look tonight.

Well, all the other v1 and v2 games run perfectly to completion with their scripts.  So only lords and snowball (both v1 and v2) have this problem with direction going out of bound (when I remove the braces to have literal text, they also run fine) ...

ajgbarnes commented:
> > Yeah I'm not surprised - I haven't exhaustively tested it but wanted it out there as "experimental". I think I can guess what's gone wrong 😄 I cannot look tonight.
> 
> Well, all the other v1 and v2 games run perfectly to completion with their scripts. So only lords and snowball (both v1 and v2) have this problem with direction going out of bound (when I remove the braces to have literal text, they also run fine) ...

I fixed this and committed to main - this was lurking in the back of my head today actually, some assumption I had back when I decompiled Snowball and Lords of Time that all the directions were always the same. It's true for cardinal directions but not beyond that. So I switched to using the dictionary and looking up the value in that.  I haven't tried all combinations yet (it's late) but Snowball and LoT work (both exits and debug).

I'll regenerate all the exits files as there will be mistakes in them because of this.

balyuf commented:
In the meantime I have ported the initial v3 changes from the decompiler to the interpreter.  It can now start a v3 game, but does not get far, since beside printing out "missing" messages, it cannot parse input, not only because dictionary support is missing, but also because the v3 input opcode works completely different (i.e. using list 9).  Remark that v3-4 also need a much larger listarea (from 512 to 2048 bytes).  With the recent addition of the debugger, it is interesting to watch the execution of v3-4 code, and see what is missing ...

So, besides v3 dictionary and messages support, mainly input, print input and get next object need to be implemented, and of course some driver functions ...

balyuf commented:
I have spent the day finding out about v3-4 dict and messages database.  It was great fun, looking at hex code, matching what the original interpreter does with it, trying to rebuild from basic understanding the Python code, checking back with the C code ...

I wish you to have the same kind of fun when you come around.  I have pushed these changes (beyond the rather trivial v3-v4 support) in a new branch, so that you can have a look at it, e.g. when you get stuck and don't want to lose too much time on it ... see https://github.com/balyuf/level9-python/tree/v3games ...

balyuf commented:
I discovered a mistake with de debugging output in the listhandler:
```
diff --git a/level9.py b/level9.py
index 1a12abb..18f0e9c 100644
--- a/level9.py
+++ b/level9.py
@@ -1264,7 +1264,7 @@ def vm_fn_listhandler(data,opCode,pc,version):
             vm_listarea[offset] = vm_variables[variable2]

         if(debugging):
-            print(f"Set list#{listNumber}[0x{variable1:02x}] = var[0x{variable2:02x}] (0x{vm_variables[variable2]:02x})")
+            print(f"Set list#{listNumber}[var[0x{variable1:02x}]] = var[0x{variable2:02x}] (0x{vm_variables[variable2]:02x})")
```

balyuf commented:
Ok, I cleanly merged your change of a few minutes ago.

I discovered a problem with exits handling in v3-v4 games where they would count on wrapping variables, e.g. subtracting below zero, would give a positive integer again, not a negative number as it currently is (not) coded.  I seem to have improved the situation, but not entirely fixed (still chasing the bug) by this change (perhaps you can think of a general way to fix this)?

```
@@ -1890,7 +1975,7 @@ def vm_fn_sub(data, opCode, pc):
     if(debugging):
         print(f"Set var[0x{secondVar:02x}] = var[0x{secondVar:02x}] (0x{vm_variables[secondVar]:04x}) - var[0x{firstVar:02x}] (0x{vm_variables[firstVar]:04x})")

-    vm_variables[secondVar] = vm_variables[secondVar] - vm_variables[firstVar]
+    vm_variables[secondVar] = (vm_variables[secondVar] - vm_variables[firstVar]) & 0xffff
 
     return pc

```

balyuf commented:
PS: in theory only the fn_add would also need this protection, as there are no other opCodes that would add or subtract?  I did change next one, but it does not solve the remaining issue(s) ... I suspect some listhandling issue (with list 7) ...

```
@@ -1946,7 +1946,7 @@ def vm_fn_add(data, opCode, pc):
     if(debugging):
         print(f"Set var[0x{secondVar:02x}] = var[0x{firstVar:02x}] (0x{vm_variables[firstVar]:04x}) + var[0x{secondVar:02x}] (0x{vm_variables[secondVar]:04x})")

-    vm_variables[secondVar] = vm_variables[firstVar] + vm_variables[secondVar]
+    vm_variables[secondVar] = (vm_variables[firstVar] + vm_variables[secondVar]) & 0xffff

     return pc

```

ajgbarnes commented:
> PS: in theory only the fn_add would also need this protection, as there are no other opCodes that would add or subtract? I did change next one, but it does not solve the remaining issue(s) ... I suspect some listhandling issue (with list 7) ...
> 
> ```
> @@ -1946,7 +1946,7 @@ def vm_fn_add(data, opCode, pc):
>      if(debugging):
>          print(f"Set var[0x{secondVar:02x}] = var[0x{firstVar:02x}] (0x{vm_variables[firstVar]:04x}) + var[0x{secondVar:02x}] (0x{vm_variables[secondVar]:04x})")
> 
> -    vm_variables[secondVar] = vm_variables[firstVar] + vm_variables[secondVar]
> +    vm_variables[secondVar] = (vm_variables[firstVar] + vm_variables[secondVar]) & 0xffff
> 
>      return pc
> ```

Yes only should be add that has this outcome, nothing else can do this, no bitshift or multiplication, all achieved through adds.   I think it's ok to do the masking but it worries me it's required.  There are certainly parts in e.g. the v1 A-code that could go past the max 16 bit value - one such case is the number of commands/turns that gets incremented each go in many of the v1 games.  You'd have to play for a loooooooong time but you could make it overflow.  There were some others but I'd have to scan the variable summaries to remind myself.  

And in an 8-bit system it doesn't matter if it overflows, just goes to the carry bit....  I did noticed though in Glenn Summers's C implementation it would overflow to the next memory location I think.  I'm not sure what say the Amiga or Archimedes engines would do.

Which game is it?  I'm worried we're hiding a problem that maybe a "feature" somewhere in my engine is creating.  Lists are the most interesting feature as I don't think the early games fully exercised all parts, and it maybe that their "architecture" defined all these great features but say only a v3 or v4 game finally used one part.  I have an unused cmdset that I was using early on to track which games used which commands / listhandlers and I may readd the addition to it.  Once I had it so it was a dictionary just to see how many times a particular operator was called - that was mind blowing.

ajgbarnes commented:
> I discovered a mistake with de debugging output in the listhandler:
> 
> ```
> diff --git a/level9.py b/level9.py
> index 1a12abb..18f0e9c 100644
> --- a/level9.py
> +++ b/level9.py
> @@ -1264,7 +1264,7 @@ def vm_fn_listhandler(data,opCode,pc,version):
>              vm_listarea[offset] = vm_variables[variable2]
> 
>          if(debugging):
> -            print(f"Set list#{listNumber}[0x{variable1:02x}] = var[0x{variable2:02x}] (0x{vm_variables[variable2]:02x})")
> +            print(f"Set list#{listNumber}[var[0x{variable1:02x}]] = var[0x{variable2:02x}] (0x{vm_variables[variable2]:02x})")
> ```

Copy and paste isn't my friend sometimes :)

balyuf commented:
> Yes only should be add that has this outcome, nothing else can do this, no bitshift or multiplication, all achieved through adds. I think it's ok to do the masking but it worries me it's required. There are certainly parts in e.g. the v1 A-code that could go past the max 16 bit value - one such case is the number of commands/turns that gets incremented each go in many of the v1 games. You'd have to play for a loooooooong time but you could make it overflow. There were some others but I'd have to scan the variable summaries to remind myself.
> 
> And in an 8-bit system it doesn't matter if it overflows, just goes to the carry bit.... I did noticed though in Glenn Summers's C implementation it would overflow to the next memory location I think. I'm not sure what say the Amiga or Archimedes engines would do.
> 
> Which game is it? I'm worried we're hiding a problem that maybe a "feature" somewhere in my engine is creating. Lists are the most interesting feature as I don't think the early games fully exercised all parts, and it maybe that their "architecture" defined all these great features but say only a v3 or v4 game finally used one part. I have an unused cmdset that I was using early on to track which games used which commands / listhandlers and I may readd the addition to it. Once I had it so it was a dictionary just to see how many times a particular operator was called - that was mind blowing.

It happens in most v3-v4 games, but I found the problem!

There is another assumption: assigning a value larger than 255 should be truncated to the lower 8-bit part.

So three things we need in general: limit sub to 16 bits, limit add to 16 bits and limit list assignment to 8 bits.  Come to think of it, in 8 bit CPUs, this will happen by itself, right?  So we have to mimic that ...  It seems logically to me, nothing to worry about ...  BTW, the GS/DK C-based interpreter does this automatically by always using the right kind of pointer (16-bit or 8-bit) ...

ajgbarnes commented:
> > Yes only should be add that has this outcome, nothing else can do this, no bitshift or multiplication, all achieved through adds. I think it's ok to do the masking but it worries me it's required. There are certainly parts in e.g. the v1 A-code that could go past the max 16 bit value - one such case is the number of commands/turns that gets incremented each go in many of the v1 games. You'd have to play for a loooooooong time but you could make it overflow. There were some others but I'd have to scan the variable summaries to remind myself.
> > And in an 8-bit system it doesn't matter if it overflows, just goes to the carry bit.... I did noticed though in Glenn Summers's C implementation it would overflow to the next memory location I think. I'm not sure what say the Amiga or Archimedes engines would do.
> > Which game is it? I'm worried we're hiding a problem that maybe a "feature" somewhere in my engine is creating. Lists are the most interesting feature as I don't think the early games fully exercised all parts, and it maybe that their "architecture" defined all these great features but say only a v3 or v4 game finally used one part. I have an unused cmdset that I was using early on to track which games used which commands / listhandlers and I may readd the addition to it. Once I had it so it was a dictionary just to see how many times a particular operator was called - that was mind blowing.
> 
> It happens in most v3-v4 games, but I found the problem!
> 
> There is another assumption: assigning a value larger than 255 should be truncated to the lower 8-bit part.
> 
> So three things we need in general: limit sub to 16 bits, limit add to 16 bits and limit list assignment to 8 bits. Come to think of it, in 8 bit CPUs, this will happen by itself, right? So we have to mimic that ... It seems logically to me, nothing to worry about ... BTW, the GS/DK C-based interpreter does this automatically by always using the right kind of pointer (16-bit or 8-bit) ...

I haven't seen a list bigger than 8-bits and yeah all variables are 16-bits.  So I agree.  And the code makes those assumptions (even Glenn Summers's).

balyuf commented:
Ok, so after coding dictionary and messages handling (including input routine) I am now up to coding the famous "get next object" functionality, which luckily is only needed for printing out descriptions of visible objects in some v3 and all v4 games.  It is not needed to finish the games, so I am tempted to first re-integrate the v3-4 input code with the existing hash and script framework, so that I can see if the games run to completion with their scripts.  Of course for the multi part v4 games to succeed, I will also minimally have to code at least the driver function to load the next part.

balyuf commented:
PS: and loading parts of mult-part games, may require us to restructure (further) the startup code, as we have to rerun a lot of it again when reinitializing the engine for the next part ...  So, I guess I will leave that for later, but there are enough v3-4 games that will not need that to already test the soundness of the implementation ...

balyuf commented:
PPS: but with python we may simply have the luxury to exec the python module again with different arguments? :-)

ajgbarnes commented:
> Ok, so after coding dictionary and messages handling (including input routine) I am now up to coding the famous "get next object" functionality, which luckily is only needed for printing out descriptions of visible objects in some v3 and all v4 games. It is not needed to finish the games, so I am tempted to first re-integrate the v3-4 input code with the existing hash and script framework, so that I can see if the games run to completion with their scripts. Of course for the multi part v4 games to succeed, I will also minimally have to code at least the driver function to load the next part.

Go ahead - I'm on my own slower journey here, I've been years at this and will pick through it at my own pace.  I'm away for the 3rd week in a row too (on an 11 hour flight right now). I wish I had more time. I'm still going to code what I code.

ajgbarnes commented:
> PS: and loading parts of mult-part games, may require us to restructure (further) the startup code, as we have to rerun a lot of it again when reinitializing the engine for the next part ... So, I guess I will leave that for later, but there are enough v3-4 games that will not need that to already test the soundness of the implementation ...

I really know little about these until I decompile some BBC games.  I want to do that anyway as I like the archeology and there are subtleties that the engine provides (get to see what they were thinking back in the day, which is wonderful).  

I have no doubt that v4 is the most challenging.  Took me a long time to understand the original BBC games (chipping away here and there).

balyuf commented:
Hi Andy,  this week I finished the v3-4 support - all games are now working fine (even the multiple choice text games, i.e. both Adrian Mole games and the Archers).  I was just trying to rebase on your changes, but I gave up.  I will leave my work as it is on the v3games branch, hopefully you can use it as a way to understand better the portable assembler code that the C-code based interpreter actually is ... ;-) if and when you come to the moment you need to implement new stuff.

I do want to mention a few more edge conditions I encountered, that might be useful to fix right away, on top of the discussion we already had with the 8 and 16 bit limitations above.  Some games use a 16 bit value to offset a list, which you then have to limit to its lower 8 bit value.  Also the value assigned itself might be a 16 bit value, to be limited to its lower 8 bit value.

Also the exits handling needs to be guarded better.  You have to take care of location 0, a too high location and a too high direction number, the first two resulting in no direct exits (inverse still possible), and the latter resulting in an inverse direction of 0xff (logical extension of the inverse table).

Finally, I do still feel interested in looking at the graphics code you have worked on.  So, if you could dump it on a branch, as it is, doesn't have to be in working order, I can try to fix, reintegrate and clean-up.  Besides that, we need to decide what to do with this PR ...

balyuf commented:
PS: your last commit (adding emerald isle) broke l9config.py and you might also want to fix the -f for loading files iso selecting games with -g, as it does not work currently (it fails on game not being defined) ...

balyuf commented:
PPS: minimal fix to get l9config and -f working again (including for v2 eden):

```
diff --git a/l9config.py b/l9config.py
index 72a22ab..90b7f74 100644
--- a/l9config.py
+++ b/l9config.py
@@ -103,7 +103,7 @@ v1Configuration = {
     "isle" : {
         "name"              : "Emerald Isle",
         "version"           : 2,
-        "filename"          : "roms/emerald.tzx"
+        "filename"          : "roms/emerald.tzx",
         "script"            : "test-scripts/emerald-isle-v2.txt",
         "locationsStartMsgId" : 0x190,
     },
diff --git a/level9.py b/level9.py
index 42ecfc7..4d16e37 100644
--- a/level9.py
+++ b/level9.py
@@ -2153,8 +2153,9 @@ def _find_v1_a_code_start(data, engineVersion):
         engineVersion = 1
     else:
         startAddress = data[0x1a] + data [0x1b] *256
+        engineVersion = 2

-    return startAddress, game
+    return startAddress, game, engineVersion


 ############################################################
@@ -2217,6 +2218,7 @@ if(args.game):
     engineVersion  = v1Configuration[game]['version']

 else:
+    game     = None
     filename = args.file
     engineVersion  = -1

@@ -2226,14 +2228,14 @@ with open(filename,'rb') as dataFile:
     dataFile.close()

 # Identify the game (do this anyway for preconfigured ones)
-aCodeStartAddr, foundGame = _find_v1_a_code_start(data, engineVersion)
+aCodeStartAddr, foundGame, engineVersion = _find_v1_a_code_start(data, engineVersion)

 # If
 if(not game):
     game = foundGame

 # If it couldn't be identified then quit
-if(aCodeStartAddr < 0 or game == 'unknown'):
+if(aCodeStartAddr < 0 and game == 'unknown'):
     print('Unable to identify the Level 9 Version 1 or Version 2 game in ' + filename)
     sys.exit(1)

```

balyuf commented:
Hi Andy, I went ahead and implemented v2-3 graphics on my v3games branch.  I closely followed the GS/DK Level 9 interpreter for the line drawing and flood filling algorithms (it was mentioned that the filling algorithm took a few iterations before fully correct).  I had no problems with it, so you can always look at how I did it (see l9pic.py).  I did not bother to integrate the graphics, just launch a (single) image viewer in which I display the (unscaled) images on Ubuntu (using eog).  I also did not implement the v4 bitmaps as there are just too many variants and besides MSX did not use them (BBC did, I believe).  Finally I decompiled and extracted all v2-3 pictures and committed them.

balyuf commented:
PS: I found the courage to also code support for v4 bitmaps (all variants).  It was more fun because of the motivation to mostly do a transliteration from C to Python (see l9bitmap.py).  I only tested it with PC version 2 bitmaps (and committed them too).  I expect it should work for BBC too, so that you would have as a reference for debugging when you come to it ...

balyuf commented:
PPS: I had fun implementing a very basic PySDL2 window iso using an external viewer (see l9graphics.py) ...

balyuf commented:
Hi Andy, in the interest of cleaning-up my fork, I will close this PR from my main branch (and force sync it to your main branch) and have my fixes on a separate branch (as should I have done to begin with).  I will then open a new PR from that new branch and leave it up to you to glean fixes from it ...
